book/src/proposals/validator-timestamp-oracle.md

```diff
@@ -0,0 +1,110 @@
+# Validator Timestamp Oracle
+
+Third-party users of Solana sometimes need to know the real-world time a block
+was produced, generally to meet compliance requirements for external auditors or
+law enforcement. This proposal describes a validator timestamp oracle that
+would allow a Solana cluster to satisfy this need.
+
+The general outline of the proposed implementation is as follows:
+
+- At regular intervals, each validator records its observed time for a known slot
+  on-chain (via a Timestamp transaction)
+- A client can request a block time using the `getBlockTime` RPC method. When a
+  client requests a timestamp for block N:
+
+  1. A validator determines a "cluster" timestamp for a recent slot before
+  block N by observing all the Timestamp transactions recorded on the
+  ledger that reference that slot, and determining the median timestamp.
```

Comment:
"median" seems inappropriate for a proof of stake cluster. How about a stake-weighted mean?